plugins/hls-notes-plugin/src/Ide/Plugin/Notes/Internal.hs

```diff
@@ -0,0 +1,89 @@
+import           Development.IDE.Core.Rules (getSourceFileSource, runAction)
+import           Ide.PluginUtils            (getNormalizedFilePath,
+                                             pluginResponse, throwPluginError)
+import           Ide.Types                  (PluginMethodHandler)
+import           Language.LSP.Types         (Method (TextDocumentDefinition),
+                                             Position (..), UInt,
+                                             type (|?) (..))
+import qualified Language.LSP.Types.Lens    as L
+import           Prelude                    hiding (lines, mod)
+import           Text.Regex.TDFA            (CompOption (..),
+                                             RegexOptions (defaultExecOpt),
+                                             defaultCompOpt, makeRegexOpts,
+                                             matchAllText)
+import           Text.Regex.TDFA.Text       (Regex)
+
+jumpToNote :: PluginMethodHandler IdeState TextDocumentDefinition
+jumpToNote state _ param = pluginResponse $ do
+    let uriOrig = param ^. (L.textDocument . L.uri)
+    nfp <- getNormalizedFilePath uriOrig
+
+    file <- liftIO $ runAction "Notes.fileContents" state (getSourceFileSource nfp)
+    let lines = BS.split '\n' file
+    (name, mod) <- case getNote (param ^. L.position) lines of
+        Just x  -> pure x
+        Nothing -> throwPluginError "No note under cursor"
+
+    let (pos, uri) = case mod of
+            Just _  -> error "not implemented"
+            Nothing -> (findNote name file (BS.length <$> lines), uriOrig)
+
+    case pos of
+        Just p  -> pure $ InL (Location uri (Range p p))
+        Nothing -> throwPluginError "Note not found"
+
+decode :: BS.ByteString -> T.Text
+decode = TE.decodeUtf8With TE.lenientDecode
+
+findNote :: T.Text -> BS.ByteString -> [Int] -> Maybe Position
+findNote name bs bslens = do
+    (pos, _) <- listToMaybe $ mapMaybe (\case
+            ((_, p):(n, _):_) | decode n == name -> Just p
+            _                                    -> Nothing
+        ) matches
+    (l, c) <- findLine bslens pos
+    pure $ Position (uint l) (uint c)
+    where
+        matches = A.elems <$> matchAllText noteRegex bs
+        uint = fromInteger . fromIntegral
+        findLine [] _ = Nothing
+        findLine (x:xs) pos | pos < x = Just (0, pos)
+                            | otherwise = first (+1) <$> findLine xs (pos - x - 1)
+
+
+getNote :: Position -> [BS.ByteString] -> Maybe (T.Text, Maybe T.Text)
+getNote (Position l c) bss = do
+    line <- decode <$> safeIndex bss l
+    let matches = A.elems <$> matchAllText noteRefRegex line
+        c' = fromInteger (fromIntegral c)
+    listToMaybe $ mapMaybe (\case
+        ((_, (s, len)):(name, _):_:(mod, _):_) -> if s <= c' && c' <= s + len then
+            Just (name, if mod == "" then Nothing else Just mod)
+            else Nothing
+        _ -> Nothing) matches
+
+noteRefRegex, noteRegex :: Regex
```

Comment:
I'd like to see some tests for these functions - tricky regexes are always annoying to reason through and easy to accidentally break!